pkg/controller/ipam/validate.go

```diff
@@ -117,141 +93,96 @@ func validationResult(allowed bool, msg string) *admv1.AdmissionResponse {
+	subnetInfo := &ipPool.Spec.SubnetInfo // IPPool requires SubnetInfo
+	ipRanges := ipPool.Spec.IPRanges
+
+	gatewayAddr, err := netip.ParseAddr(subnetInfo.Gateway)
+	if err != nil {
+		return fmt.Sprintf("invalid gateway address %s", subnetInfo.Gateway), false
 	}
 
-	var difference []crdv1beta1.IPRange
-	for _, ipRange := range s1 {
-		if _, exists := newSet[ipRange]; exists {
-			continue
+	if gatewayAddr.Is4() {
+		if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 32 {
+			return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+		}
+	} else {
+		if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
+			return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
 		}
-		difference = append(difference, ipRange)
 	}
-	return difference
-}
+	prefix := netip.PrefixFrom(gatewayAddr, int(subnetInfo.PrefixLength)).Masked()
+	subnet := &prefix
 
-func humanReadableIPRanges(ipRanges []crdv1beta1.IPRange) string {
-	strs := make([]string, len(ipRanges))
-	for i, ipRange := range ipRanges {
-		if ipRange.CIDR != "" {
-			strs[i] = ipRange.CIDR
-		} else {
-			strs[i] = fmt.Sprintf("%s-%s", ipRange.Start, ipRange.End)
+	combinedRanges := make(map[string][2]netip.Addr)
+	for _, pool := range existingIPPools {
+		if pool.Name == ipPool.Name {
+			continue
+		}
+		for _, ipRange := range pool.Spec.IPRanges {
+			var key string
+			var start, end netip.Addr
+
+			if ipRange.CIDR != "" {
+				key = fmt.Sprintf("range [%s] of pool %s", ipRange.CIDR, pool.Name)
+				cidr, _ := validation.ParseIPRangeCIDR(ipRange.CIDR)
+				start, end = utilip.GetStartAndEndOfPrefix(cidr)
+			} else {
+				key = fmt.Sprintf("range [%s-%s] of pool %s", ipRange.Start, ipRange.End, pool.Name)
+				start, end, _ = validation.ParseIPRangeStartEnd(ipRange.Start, ipRange.End)
+			}
+			combinedRanges[key] = [2]netip.Addr{start, end}
 		}
 	}
-	return fmt.Sprintf("[%s]", strings.Join(strs, ","))
-}
 
-func rangesOverlap(r1, r2 crdv1beta1.IPRange) bool {
-	if r1.CIDR == "" {
-		if r2.CIDR == "" {
-			r1start := net.ParseIP(r1.Start)
-			r1end := net.ParseIP(r1.End)
-			r2start := net.ParseIP(r2.Start)
-			r2end := net.ParseIP(r2.End)
+	for _, ipRange := range ipRanges {
+		var key string
+		var start, end netip.Addr
 
-			if ipInRange(r1start, r1end, r2start) || ipInRange(r2start, r2end, r1start) {
-				return true
+		if ipRange.CIDR != "" {
+			key = fmt.Sprintf("range [%s]", ipRange.CIDR)
+			cidr, errMsg := validation.ParseIPRangeCIDR(ipRange.CIDR)
+			if errMsg != "" {
+				return errMsg, false
 			}
-		} else {
-			_, cidr2, _ := net.ParseCIDR(r2.CIDR)
-			r1start := net.ParseIP(r1.Start)
-			r1end := net.ParseIP(r1.End)
-			if cidr2.Contains(r1start) || cidr2.Contains(r1end) {
-				return true
+			start, end = utilip.GetStartAndEndOfPrefix(cidr)
+			if (!gatewayAddr.Is4() || !start.Is4() || !end.Is4()) && (!gatewayAddr.Is6() || !start.Is6() || !end.Is6()) {
```

Comment:
I am not sure this expression is correct. Do not quite understand its logic. Could we do: `start.Is4() != gatewayAddr.Is4() || end.Is4() != gatewayAddr.Is4()`

Reply:
Good idea. Updated